User controlled havoc(ing) for memory-snapshot harness

xbauch opened the pull request:
In #4438 we build support for recovering structs and pointers in goto-harness from a memory snapshot. The havocing there was by default: the user couldn't change how deeply will the structs and array-pointers initialised. This PR unifies the control over non-deterministic initialisation between function-call and memory-snapshot harnesses, consequently allowing the user to control the initialisation in memory-snapshot harness as well.

Pointers to arrays are initialised in a similar fashion to function-call harness, i.e. the user has to declare which variable contains the size of the array.

Also dynamically allocated pointers in the source-program and changed into statically allocated arrays in the memory snapshot, so that it would be legal to query about their content in the harness-program.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

sonodtt commented on src/goto-harness/recursive_initialization.h:
```diff
@@ -35,6 +38,10 @@ struct recursive_initialization_configt
 
   std::string to_string() const; // for debugging purposes
+
+  bool handle_option(
```
Descriptive comment please.

xbauch replied:
Done.

sonodtt commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -415,3 +416,56 @@ recursive_initializationt::cstring_member_initialization()
   };
 }
+
+bool recursive_initialization_configt::handle_option(
```
Can be moved to top of file? (preserve expected ordering)

xbauch replied:
Done.

sonodtt commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -90,57 +90,19 @@ void function_call_harness_generatort::handle_option(
+  auto &require_exactly_one_value =
+    harness_options_parsert::require_exactly_one_value;
+
+  if(p_impl->recursive_initialization_config.handle_option(option, values))
```
empty branch - remind us of the why and what is going on here?
People will typically be viewing the code in a hurry, and not in an ordered PR as here. This may be where they jump in.

xbauch replied:
Done.

sonodtt commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -222,7 +226,7 @@ code_blockt memory_snapshot_harness_generatort::add_assignments_to_globals(
 {
   recursive_initializationt recursive_initialization{
-    recursive_initialization_configt{}, goto_model};
+    recursive_initialization_config, goto_model};
```
Extraneous removal of brace initialization that clarifies this is a temporary object constructed in place (and initialised via value initialization).
Not sure you gain additional clarity by removing it.

xbauch replied:
It's not a temporary object anymore.

sonodtt replied:
Ah! Good.

sonodtt commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -23,18 +25,20 @@ Author: Daniel Poetzl
-  if(option == "memory-snapshot")
+  auto &require_exactly_one_value =
+    harness_options_parsert::require_exactly_one_value;
+  if(recursive_initialization_config.handle_option(option, values))
```
empty branch - remind us of the why and what is going on here?
People will typically be viewing the code in a hurry, and not in an ordered PR as here. This may be where they jump in.  (same remark in function_call_harness_generator.cpp)

xbauch replied:
Done.

sonodtt commented on regression/snapshot-harness/nondet_initialize_static_arrays/main.c:
```diff
@@ -0,0 +1,51 @@
+  checkpoint();
+
+  assert(p != NULL);
+  for(int index = 0; index < 3; index++)
```
Comment. Yes - its clear, but for this may be where future someone starts in figuring out the code.

This principle holds elsewhere in the tests.

xbauch replied:
Maybe done.

sonodtt commented on regression/snapshot-harness/nondet_initialize_static_arrays/main.c:
```diff
@@ -0,0 +1,51 @@
+    assert(p->array[index]->to_st2->to_st1->to_st2 == NULL);
+  }
+
+  assert(p->array[0] != p->array[1]);
```
Definitely some comments here. Clarify why the != holds

This principle holds elsewhere in the tests.

xbauch replied:
Maybe done.

sonodtt commented on regression/snapshot-harness/nondet_initialize_static_arrays/main.c:
```diff
@@ -0,0 +1,51 @@
+
+st1_t dummy1;
+st2_t dummy2;
+
```
Hmm. The dummys are not needed in this test? Remove anything extraneous or we'll be wondering if we've overlooked something.

xbauch replied:
Removed.

sonodtt commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -190,7 +190,8 @@ exprt symbol_analyzert::get_char_pointer_value(
   else
   {
-    return it->second;
+    return symbol_exprt{to_symbol_expr(it->second).get_identifier(),
```
comment? if you had to understand something - please tell us what and why. 
This should greatly help in reducing review time and increase accuracy of reviews.

xbauch replied:
Done, or maybe removed.

sonodtt commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -315,7 +316,16 @@ exprt symbol_analyzert::get_pointer_value(
+        "pointers-to-char shouldn't point to members");
+      const auto result_pointee_expr =
+        get_char_pointer_value(expr, memory_location, location);
+      const auto maybe_cast_result = typecast_exprt::conditional_cast(
```
hmm. My guess is you had to understand something here about memory analyser. 
Any clarification would greatly help reviewers. This code was inherited, but as we understand it (to extend it) - we should clarify, so as to not duplicate effort when this code is passed on to others.

xbauch replied:
Done, or maybe rewritten.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 197f23db43746b5156d43f4fc1c1e85cfb49090e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/109281843
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

tautschnig commented on src/goto-harness/goto_harness_generator.h:
```diff
@@ -16,6 +16,27 @@ Author: Diffblue Ltd.
 
 class goto_modelt;
 
+class harness_options_parsert
```
Hmm, so this really is using a class as a namespace, just because our coding guidelines say so?

xbauch replied:
It was not the intention, but I guess that is the only reason.

tautschnig replied:
Then maybe we should actually use a namespace?

xbauch replied:
Here: https://github.com/diffblue/cbmc/pull/4482/commits/69811504a82c777d126ef3da318c1e01e232e3c2. I'm just not sure about the naming convention: should a namespace end with `t` or not?

xbauch replied:
Also the linter doesn't like it very much.

tautschnig replied:
I think you fixed the latter? Looks good to me. But I'm not sure what others think about it :-)

xbauch replied:
It was suggested that rather than rewriting the linter we should just wrap the namespace in `// NOLINT` comments: https://github.com/diffblue/cbmc/pull/4482/commits/b9fd0afe4aec9fe83bd607f1c69fed94135c322b.

tautschnig commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -16,9 +16,63 @@ Author: Diffblue Ltd.
 #include <util/optional_utils.h>
 #include <util/std_code.h>
 #include <util/std_expr.h>
+#include <util/string2int.h>
```
The commit message sounds cool, but really doesn't give my any information I'm afraid. Please clarify.

xbauch replied:
Sorry about the sloppy description, expanded here: https://github.com/diffblue/cbmc/pull/4482/commits/d96545dc3db54af5bed3d7d66dbc69efd48db6bd.

tautschnig commented on src/memory-analyzer/gdb_api.h:
```diff
@@ -83,6 +83,12 @@ class gdb_apit
     const optionalt<std::string> string;
   };
 
+  /// Get the allocated size estimate for a pointer
```
Could you explain why this is an estimate? And is it always either an under- or an overestimate?

xbauch replied:
Added explanation: https://github.com/diffblue/cbmc/pull/4482/commits/ef457751a2867d1cb29b918ec98a2680bad80b24.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -341,28 +341,51 @@ exprt gdb_value_extractort::get_pointer_value(
-      return get_char_pointer_value(expr, memory_location, location);
+      const auto target_expr =
+        get_pointer_to_member_value(expr, value, location);
+      CHECK_RETURN(target_expr.id() != ID_nil);
```
I'd prefer `target_expr.is_not_nil()`

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -341,28 +341,51 @@ exprt gdb_value_extractort::get_pointer_value(
+        : get_non_char_pointer_value(expr, memory_location, location);
+
+    // postpone if we cannot resolve now
+    if(target_expr.id() == ID_nil)
```
`if(target_expr.is_nil())`

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -247,6 +247,29 @@ exprt gdb_value_extractort::get_pointer_to_member_value(
+    const auto val = get_expr_value(symbol_expr, *zero, location);
+
+    symbol_exprt dummy(
+      pointer_typet(symbol_expr.type(), config.ansi_c.pointer_width));
```
Use `pointer_type` from `util/c_types.{h,cpp}`, and add a name.

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -300,18 +300,61 @@ exprt gdb_value_extractort::get_non_char_pointer_value(
+
+      // allocate a new symbol for the temporary static array
+      symbol_exprt array_dummy(
+        pointer_typet(target_array_type, config.ansi_c.pointer_width));
```
Use `pointer_type`, add a name.

xbauch replied:
Done.

petr-bauch committed:
Refactor and unify harness generator options

Extract the common options and create options file for the memory-snapshot
specific ones. Also change one description and add pointers-as-strings
description of function harness.

petr-bauch committed:
Extract static option helpers into a separate class

To be used also for recursive init config type.

petr-bauch committed:
Handle the config specific options in-house

via a member method handle_option. That is recursive_initialization_config now
implements its own handle_option that parses the user options specific to the
recursive initialization. These are: minimal null tree depth, maximal
nondeterministic tree depth, maximal array size and minimal array size. To that
end, the config inherits from the harness_option_parser, e.g. to have access to
require_exactly_one_value, etc.

petr-bauch committed:
Reflect the changes on individual generators

1: Config options are handled by the config's internal `handle_option`.
2: Options names are unified.
3: Memory-snapshot options are in a separate file.
4: Fix parsing the list of variables to be havoced.
5: Parse the treat-pointers-as-arrays in memory snapshot.

tautschnig commented:
@xbauch This now needs a rebase.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: df538f03d3a7d930910598a8474487bbe68acbc3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112133832

tautschnig commented on src/goto-harness/function_harness_generator_options.h:
```diff
@@ -9,13 +9,10 @@ Author: Diffblue Ltd.
-  "max-nondet-tree-depth"
-#define FUNCTION_HARNESS_GENERATOR_MIN_ARRAY_SIZE_OPT "min-array-size"
-#define FUNCTION_HARNESS_GENERATOR_MAX_ARRAY_SIZE_OPT "max-array-size"
+#define COMMON_HARNESS_GENERATOR_NONDET_GLOBALS_OPT "nondet-globals"
```
It seems surprising that this option isn't in `common_harness_generator_options.h`?!

xbauch replied:
Sorry, my bad. In fact, it's the other way around. This option is `function-harness` specific and will be renamed.

tautschnig commented on src/goto-harness/recursive_initialization.h:
```diff
@@ -35,6 +38,15 @@ struct recursive_initialization_configt
+  /// \param option: the user option name
+  /// \param values: the (one-or-more) values for this option
+  /// \return true if the option belonged to recursive initialisation and was
+  /// successfully parsed here
```
Nit pick: indent comment that spans multiple lines (as documented in CODING_STANDARD.md)

xbauch replied:
Done.

tautschnig commented on unit/memory-analyzer/gdb_api.cpp:
```diff
@@ -167,9 +167,16 @@ TEST_CASE("gdb api test", "[core][memory-analyzer]")
-    REQUIRE(gdb_api.get_value("z") == "c");
+    const auto &x_value = gdb_api.get_value("x");
+    const auto &y_value = gdb_api.get_value("y");
+    const auto &z_value = gdb_api.get_value("z");
```
I'd prefer if changes of this kind were part of the commit that changes the API. That will make sure that bisecting remains possible.

xbauch replied:
Ok, I'll squash it together with the change that caused it.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b30bb658c6de0739e1c3e70479b5102a887978d6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112204059

NlightNFotis reviewed (approved):
Looks good to me.

chrisr-diffblue commented on regression/snapshot-harness/pointer-to-array-char/main.c:
```diff
@@ -0,0 +1,29 @@
+
+  assert(first == second);
+  assert(second[array_size - 1] == 'a');
+  //  assert(second[10]=='0');
```
Is there a reason this assert is still commented out? If so, maybe worth including a comment explaining why.

xbauch replied:
This one is redundant. Removed.

chrisr-diffblue commented on regression/snapshot-harness/pointer-to-array-int/main.c:
```diff
@@ -0,0 +1,42 @@
+  checkpoint();
+
+  assert(first == second);
+  /* assert(array_size >= prefix_size); */
```
Ditto with these asserts - worth explaining why they don't currently hold/when they might be able to be enabled.

xbauch replied:
This one was postponed until the improved support for dynamic allocation will be implemented in #4578. I've added a comment.

chrisr-diffblue commented on src/memory-analyzer/gdb_api.cpp:
```diff
@@ -56,6 +57,39 @@ gdb_apit::~gdb_apit()
 
+size_t gdb_apit::query_malloc_size(const std::string &pointer_expr)
+{
+  write_to_gdb("-var-create tmp * malloc_usable_size(" + pointer_expr + ")");
```
I believe `malloc_usable_size` is only available on Linux. On OSX I think the equivalent function on Mac OS X is `malloc_size` - so we probably need to do some conditional compilation guarding here...

xbauch replied:
I thought the machine interface worked so much differently under OSX that we only wanted to compile there but not actually run the `memory-analyzer`. Anyway what do you think about https://github.com/diffblue/cbmc/pull/4482/commits/57f08e14344316b19c9ae5ecc828d241f28d27be?

chrisr-diffblue reviewed (changes requested):
Mostly just nits/queries - but I think the unconditional use of `malloc_useable_size` needs addressing for future sanity.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 47138b752e39a5a810a6513f1964d4de10cb7761).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112282432
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 57f08e14344316b19c9ae5ecc828d241f28d27be).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112284250

chrisr-diffblue reviewed (approved):
Looks good to me now, thanks!

petr-bauch committed:
Add a method to get the malloc(ed) size estimate

Returning 1 if variable was not malloc(ed). The estimation is described in the
documentation.

petr-bauch committed:
Optionalize the results of querying gdb

Not all queried values are sensible if we ask past allocated
memory (malloc_usable_size doesn't return exact values).

petr-bauch committed:
Sort the snapshot symbols based on pointer-depth

To prevent assigning uninitialized values. Also build the
recursive_initialization with the member config.

petr-bauch committed:
Refactor snapshot-harness interface

1: Harness option parser is a namespace not a class: its methods are not static
   and we do not inherit from it.
2: Rename mis-labelled option: from common-harness..-nondet-globals to
   function-harness-..-nondet-globals.

petr-bauch committed:
Clarify expected types of expressions

Also pointers to members can be chars.

petr-bauch committed:
Avoid converting to string

in id2boolean.

petr-bauch committed:
Clear get_pointer_value

Unify handling pointers-to-members/to-char/to-non-char and add comments.

petr-bauch committed:
Add building temporary object for pointer-to-member

Using the type from symbol table and value from gdb.

petr-bauch committed:
Handle the dynamically allocated pointers

By pretending they were allocated statically. We get the estimated allocated
size and build a static object of that size.

petr-bauch committed:
Regression tests for user-control of havoc

Testing that the user can set how deeply a struct/pointer is
non-deterministically initialized as well as some cases regarding pointers,
strings, dynamic allocation, etc.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: e0d3e5e7c262bf9e7e159bfd87588a5e278a4882).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112430842
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
